lib/Sema/CSApply.cpp

```diff
@@ -2862,11 +2862,67 @@ namespace {
       // Check for ambiguous member if the base is an Optional
       if (baseTy->getOptionalObjectType()) {
         diagnoseAmbiguousNominalMember(baseTy, result);
+      } else {
+        tryDiagnoseContextualOptionalNoneMemberAmbiguous(expr, baseTy);
       }
 
       return coerceToType(result, resultTy, cs.getConstraintLocator(expr));
     }
-    
+
+    /// Diagnose an unresolved `.none` member infered from a contextual optional
+    /// type where the member base type is infered as a static non-optional member.
+    /// So we enforce explicit type anotation to avoid ambiguity.
+    ///   enum Enum<T> {
+    ///     case bar
+    ///     static var none: Enum<Int> { .bar }
+    ///   }
+    ///   let _: Enum? = .none // Infered as static member Enum<Int>.none, emit
+    ///   warning suggesting explicit type.
+    ///   let _: Enum = .none // Ok
+    ///
+    void tryDiagnoseContextualOptionalNoneMemberAmbiguous(UnresolvedMemberExpr *UME,
+                                                          Type baseTy) {
+      assert(!baseTy->getOptionalObjectType());
+      
+      if (!UME->getName().isSimpleName("none"))
```

Comment:
The converse diagnostics below seem to be agnostic to whether the ambiguous member is `.some` or `.none`, but I don't actually see any tests for the `.some` condition... should we check for `"some"` as well here?

Reply:
I think it does ignores it in some way because the diagnostic is not produced for the none case:
```swift
struct Foo<T> {
  static var none: Foo<Int> { Foo<Int>() }
  static func some(_: Foo<Int>) -> Foo<Int> { Foo<Int>() }
}

let _: Foo<Int>? = .some(Foo<Int>()) // Optional.some
let _: Foo? = .some(Foo()) // Foo<Int>.some static member
let _: Foo? = .some(Foo<Int>()) // Optional.some
let _: Foo = .some(Foo<Int>())  // Foo<Int>.some static member
```